metrics/metrics.go

```diff
@@ -101,26 +208,105 @@ var _ CSIMetricsManager = &csiMetricsManager{}
-// operationErr - Error, if any, that resulted from execution of operation.
-// operationDuration - time it took for the operation to complete
+// RecordMetrics implements CSIMetricsManager.RecordMetrics.
 func (cmm *csiMetricsManager) RecordMetrics(
 	operationName string,
 	operationErr error,
 	operationDuration time.Duration) {
-	cmm.csiOperationsLatencyMetric.WithLabelValues(
-		cmm.driverName, operationName, getErrorCode(operationErr)).Observe(operationDuration.Seconds())
+	cmm.recordMetricsWithLabels(operationName, operationErr, operationDuration)
+}
+
+// recordMetricsWithLabels is the internal implementation of RecordMetrics.
+func (cmm *csiMetricsManager) recordMetricsWithLabels(
+	operationName string,
+	operationErr error,
+	operationDuration time.Duration,
+	labelValues ...string) {
+	values := []string{cmm.driverName, operationName, getErrorCode(operationErr)}
+	toAdd := len(labelValues)
+	if toAdd > len(cmm.additionalLabelNames) {
+		// To many labels?! Truncate. Shouldn't happen because of
+		// error checking in WithLabelValues.
+		toAdd = len(cmm.additionalLabelNames)
+	}
+	values = append(values, labelValues[0:toAdd]...)
+	for i := toAdd; i < len(cmm.additionalLabelNames); i++ {
+		// Backfill missing values with empty string.
+		values = append(values, "")
+	}
+	for _, label := range cmm.additionalLabels {
+		values = append(values, label.value)
+	}
+	cmm.csiOperationsLatencyMetric.WithLabelValues(values...).Observe(operationDuration.Seconds())
+}
+
+type csiMetricsManagerWithValues struct {
+	*csiMetricsManager
+
+	// additionalValues holds the values passed via WithLabelValues.
+	additionalValues []string
+}
+
+// WithLabelValues in the base metrics manager creates a fresh wrapper with no labels and let's
+// that deal with adding the label values.
+func (cmm *csiMetricsManager) WithLabelValues(labels map[string]string) (CSIMetricsManager, error) {
+	cmmv := &csiMetricsManagerWithValues{csiMetricsManager: cmm}
+	return cmmv.WithLabelValues(labels)
+}
+
+// WithLabelValues in the wrapper creates a wrapper which has all existing labels and
+// adds the new ones, with error checking.
+func (cmmv *csiMetricsManagerWithValues) WithLabelValues(labels map[string]string) (CSIMetricsManager, error) {
+	extended := &csiMetricsManagerWithValues{cmmv.csiMetricsManager, append([]string{}, cmmv.additionalValues...)}
+
+	for name := range labels {
+		if !cmmv.haveAdditionalLabel(name) {
+			return nil, fmt.Errorf("label %q was not defined via WithLabelNames", name)
+		}
+	}
+	// Add in same order as in the label definition.
+	for _, name := range cmmv.additionalLabelNames {
+		if value, ok := labels[name]; ok {
+			if len(cmmv.additionalValues) >= len(extended.additionalLabelNames) {
+				return nil, fmt.Errorf("label %q = %q cannot be added, all labels already have values %v", name, value, cmmv.additionalValues)
+			}
+			extended.additionalValues = append(extended.additionalValues, value)
+		}
```

Comment:
I think we should return an error if the value is not specified. Otherwise all your subsequent values will get offset from the label.

Reply:
The intention was to mirror the semantic of https://godoc.org/github.com/prometheus/client_golang/prometheus#SummaryVec.CurryWith which allows adding values repeatedly. I've made that clearer in the function comment.